rootston/cursor.c

```diff
@@ -98,51 +99,56 @@ static void seat_view_deco_button(struct roots_seat_view *view, double sx,
 	}
 }
 
-static void roots_cursor_update_position(struct roots_cursor *cursor,
+static void roots_passthrough_cursor(struct roots_cursor *cursor,
 		uint32_t time) {
-	struct roots_desktop *desktop = cursor->seat->input->server->desktop;
-	struct roots_seat *seat = cursor->seat;
 	struct roots_view *view;
-	struct wlr_surface *surface = NULL;
 	double sx, sy;
-	switch (cursor->mode) {
-	case ROOTS_CURSOR_PASSTHROUGH:
-		view = desktop_view_at(desktop, cursor->cursor->x, cursor->cursor->y,
-			&surface, &sx, &sy);
+
+	struct roots_seat *seat = cursor->seat;
+	struct roots_desktop *desktop = seat->input->server->desktop;
+	struct wlr_surface *surface = desktop_surface_at(desktop,
+			cursor->cursor->x, cursor->cursor->y, &sx, &sy, &view);
+
+	if (!surface && cursor->cursor_client) {
+		wlr_xcursor_manager_set_cursor_image(cursor->xcursor_manager,
+			cursor->default_xcursor, cursor->cursor);
+		cursor->cursor_client = NULL;
+	}
+
+	if (view) {
 		struct roots_seat_view *seat_view =
 			roots_seat_view_from_view(seat, view);
-		if (cursor->pointer_view && (surface || seat_view != cursor->pointer_view)) {
+		if (cursor->pointer_view && (surface ||
+					seat_view != cursor->pointer_view)) {
 			seat_view_deco_leave(cursor->pointer_view);
 			cursor->pointer_view = NULL;
 		}
-		bool set_compositor_cursor = !view && !surface && cursor->cursor_client;
-		if (view && surface) {
-			struct wl_client *view_client =
-				wl_resource_get_client(view->wlr_surface->resource);
-			set_compositor_cursor = view_client != cursor->cursor_client;
-		}
-		if (set_compositor_cursor) {
-			wlr_xcursor_manager_set_cursor_image(cursor->xcursor_manager,
-				cursor->default_xcursor, cursor->cursor);
-			cursor->cursor_client = NULL;
-		}
-		if (view && !surface) {
-			if (seat_view) {
-				cursor->pointer_view = seat_view;
-				seat_view_deco_motion(seat_view, sx, sy);
-			}
-		} if (view && surface) {
-			// motion over a view surface
-			wlr_seat_pointer_notify_enter(seat->seat, surface, sx, sy);
-			wlr_seat_pointer_notify_motion(seat->seat, time, sx, sy);
-		} else {
-			wlr_seat_pointer_clear_focus(seat->seat);
+		if (!surface) {
+			cursor->pointer_view = seat_view;
+			seat_view_deco_motion(seat_view, sx, sy);
 		}
+	}
 
-		struct roots_drag_icon *drag_icon;
-		wl_list_for_each(drag_icon, &seat->drag_icons, link) {
-			roots_drag_icon_update_position(drag_icon);
-		}
+	if (surface) {
```

Comment:
Should set the default cursor if `cursor->cursor_client != wl_resource_get_client(surface->resource)`